src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs

```diff
@@ -309,8 +326,21 @@ private ILEmitResolverBuilderRuntimeContext GenerateMethodBody(ServiceCallSite c
 
                 Label skipCreationLabel = context.Generator.DefineLabel();
                 Label returnLabel = context.Generator.DefineLabel();
+                Label defaultLabel = context.Generator.DefineLabel();
+
+                // Check if scope IsRootScope
+                context.Generator.Emit(OpCodes.Ldarg_1);
+                context.Generator.Emit(OpCodes.Callvirt, ScopeIsRootScope);
+                context.Generator.Emit(OpCodes.Brfalse_S, defaultLabel);
+
+                context.Generator.Emit(OpCodes.Call, CallSiteRuntimeResolverInstanceField);
+                AddConstant(context, callSite);
```

Comment:
This is a nice change. @pakrym but we're paying for bounds checks!! 😄

Reply:
We can change AddConstant to produce unsafe code with `fixed`.

Reply:
Don't think it matters though.

Reply:
This method is locking and has a try finally so I'd be surprised if it mattered as well.

> We can change AddConstant to produce unsafe code with fixed.

`Unsafe.Add`